### PR TITLE
pull.without-history - smaller pulls

### LINKED ISSUE #2696: Support squashing a branch as part of a `pull`.
Goal: Reduce codebase sizes by avoiding pulling in outdated history and references which aren't present in the HEAD of a namespace.

Problem: 
Currently, even if you `pull`, and then immediately `squash`, the history isn't garbage-collected, and we currently don't have an easy way to clean orphaned branches or references. This means they're likely to hang out in your codebase and artificially inflate its size.

Solution:
* If we squash the branch in-memory before we sync it to the codebase we avoid adding the orphaned definitions in the first place.

---

Acceptance Criteria
- [x]  Allow pulling a remote branch into your local codebase without all of its history.
- [x] This must happen before syncing, because that will inflate the local codebase.

### PR DESCRIPTION
**Pull request by ChrisPenner**
This was added for use by the git-hosted Unison Share transcript, primarily to shrink the file and download size for Share hosted on Github. Note that this command doesn't propagate any patch.
___

## Overview

Adds a new variant of the `pull` command which discards the history of a remote branch **before** importing it.
This is important, once a branch and its history is pulled in, all of its definitions throughout all of its history will be part of your local codebase, and even if you squash the branch afterwards those definitions will be floating around your codebase until a manual garbage-collect or migration step.

This avoids that problem by squashing the branch in memory BEFORE importing it,  meaning you only pay the storage cost for the HEAD of that branch.

I switched to use this mode for the `base` pull on startup as well.

Results:

```
pull.without-history https://github.com/unisonweb/base .base
- 37.7s, 6mb codebase

pull https://github.com/unisonweb/base .base
- 85.7s (some significant amount of this is likely in constructing the diff), 14mb codebase
```

closes #2696

## Implementation notes

* I add a new `pull.without-history` command
* It works by passing a **preprocessing** step to `ImportRemoteBranch`
* I swap the pull-base in the onboarding flow to use a `pull.without-history`

## Interesting/controversial decisions

The addition of this command further makes the distinction between code that you're **working on** and code that you only **depend on**. I think projects will make this more clear in the future, but for now we may want to think about what sort of language we use to refer to these things.

I'm not sold on the name `pull.without-history`, after some chats with @hojberg  we agreed that this should probably be a "default" workflow when pulling in dependencies.

Options we've thought of for names:
* `load`
* `install`
* Swap this to be `pull` and add `pull.with-history`
* ??? (your name here)

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase.hs]: The shadowing is intentional BTW.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1684]: We don't propagate the patch here since we take the remote snapshot as-is.

My understanding is that the patch is only propagated _within the merged result_, e.g. if I merge `remote:.base` into `.base` it will propagate the patch within `.base` after merging.

This is unnecessary `.base` is always an exact copy of `remote:.base` as it is in `pull.without-history`.

Let me know if I'm missing something here.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: I ran ormolu over the imports.

File: unison-cli/src/Unison/CommandLine/Welcome.hs
Comment (ChrisPenner) [inline unison-cli/src/Unison/CommandLine/Welcome.hs]: Pull base without history as part of on-boarding, it's faster and cuts down on codebase size.

If folks want to develop on base they need to pull trunk anyways.

**Review by pchiusano — APPROVED** [review]
Sweet!

Name seems good to me. I think we'll have some higher level commands eventually for things like library installation / upgrading / etc, and don't think `pull` will be in the name.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1684]: I didn't quite follow. Is this pull guaranteed to not be merging anything? i.e. it's rejected for non-empty namespaces?

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Branch.hs]: I should've asked in the previous PR, but since we're working here again: Is there a good reason to have these arguments in the reverse of the normal order for `cons`?

File: unison-cli/src/Unison/CommandLine/Welcome.hs
Comment (aryairani) [reply unison-cli/src/Unison/CommandLine/Welcome.hs]: Let's chat about this, I don't think it's the right approach for slimming base.

Maybe there's a more general preprocessing trick, like `pure . Branch.getAt path` rather than `pure . Branch.discardHistory`?  Or if you wanna get the core of this PR merged and improve the base stuff in a subsequent PR, that's okay too.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: If you do that, go ahead and move the `{- ORMOLU_DISABLE -}` below the imports for the future.

**Review by aryairani — COMMENTED** [review]
Looking good! I left a couple of comments to discuss.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (ChrisPenner) [reply unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: I've tried that and it doesn't work on my machine, but if it works on yours I can do that.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (ChrisPenner) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1684]: Currently, `pull.without-history remote local` conses a 'snapshot' of the remote's HEAD onto the local history.
So no, it won't ever **merge**, if you want merges, you're probably working on this project and you need to pull with history.
Think of pull.without-history like you're pulling a dependency of your project, if I upgrade from `random` version 1 to 1.5, I don't care about keeping the 'history' of `random`, I just want the latest version.

You will want to apply the patch after pulling the latest version in most cases though.
What I'm asking about is whether that's meant to happen automatically and AFAICT the answer is no.

File: parser-typechecker/src/Unison/Codebase/Branch.hs
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Branch.hs]: I suppose I think of history as going from left-to-right, so 'cons-ing' in history seems more like a `snoc`; I see your point though , I don't mind changing it.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: Oh, no, I haven't tried it. Do you mean that when you tried moving `{- ORMOLU_DISABLE -}` below the imports, ormolu still didn't format the imports? Or what didn't work?

File: unison-cli/src/Unison/CommandLine/Welcome.hs
Comment (ChrisPenner) [reply unison-cli/src/Unison/CommandLine/Welcome.hs]: I reverted it here, I'm still of the opinion we should turn this on, but we can discuss this elsewhere.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (ChrisPenner) [reply unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: > Do you mean that when you tried moving {- ORMOLU_DISABLE -} below the imports, ormolu still didn't format the imports?

Yes, that's what I observed, but didn't dig into why that's the case (it didn't seem super important)

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1684]: Ok, I think I get it.

File: unison-cli/src/Unison/Codebase/Editor/Input.hs
Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/Input.hs:2]: Ok, I opened #2746 to track.

**Review by aryairani — APPROVED** [review]
Ok, looks good!  One final thought (you can have the final say) wdyt about an ADT for the preprocessing step, like

```
data Preprocessing m = Unmodified | Preprocessed (Branch m -> m (Branch m))
```
where `Unmodified` might be a little more obviously interpretable at the call sight.

**Pull request merged**
